bpf/protocol_common.h

```diff
@@ -95,114 +97,189 @@ static __always_inline http_connection_metadata_t *connection_meta_by_direction(
+        ctx->nr_segs = iter.nr_segs;
+        ctx->iov = iter.iov;
+        ctx->ubuf = iter.ubuf;
     } else {
-        // TODO: I wonder if there's a way to check for field existence without having to
-        // make fake structures that match the new version of the kernel code. This code
-        // here assumes the kernel iov_iter structure is the format with __iov and __ubuf_iovec.
-        struct _iov_iter _msg_iter;
-        bpf_probe_read_kernel(&_msg_iter, sizeof(struct _iov_iter), &(msg->msg_iter));
-
-        bpf_dbg_printk("new kernel, iov doesn't exist, nr_segs %d", _msg_iter.nr_segs);
-        if (msg_iter_type == 5) {
-            struct iovec vec;
-            bpf_probe_read(&vec, sizeof(struct iovec), &(_msg_iter.__ubuf_iovec));
-            bpf_dbg_printk("ubuf base %llx, &ubuf base %llx", vec.iov_base, &vec.iov_base);
-
-            bpf_probe_read(buf, l, vec.iov_base);
-            return l;
-        } else {
-            bpf_probe_read(&iov, sizeof(struct iovec *), &(_msg_iter.__iov));
-        }     
+        struct iov_iter___v64 iter;
+        bpf_core_read(&iter, sizeof(iter), &msg->msg_iter);
+
+        ctx->iter_type = iter.iter_type & 0xff;
+        ctx->iov_offset = iter.iov_offset;
+        ctx->count = iter.count;
+        ctx->nr_segs = iter.nr_segs;
+        ctx->iov = iter.__iov;
+        ctx->ubuf = iter.ubuf;
     }
-    
-    if (!iov) {
+}
+
+static __always_inline int read_msghdr_buf(struct msghdr *msg, u8* buf, size_t max_len) {
+    if (max_len == 0) {
         return 0;
     }
 
-    if (msg_iter_type == 6) {// Direct char buffer
-        bpf_dbg_printk("direct char buffer type=6 iov %llx", iov);
-        bpf_probe_read(buf, l, iov);
+    bpf_clamp_umax(max_len, IO_VEC_MAX_LEN);
 
-        return l;
+    struct iovec_iter_ctx ctx;
+
+    get_iovec_ctx(&ctx, msg);
+
+    const int iter_ubuf = LINUX_KERNEL_VERSION > KERNEL_VERSION(6, 7, 0) ? 0 : 6;
+
+    bpf_printk("t=%u, off=%llu, count=%llu", ctx.iter_type, ctx.iov_offset, ctx.count);
+    bpf_printk("nr_segs=%lu, iov=%p, ubuf=%p", ctx.nr_segs, ctx.iov, ctx.ubuf);
+
+    if (ctx.count == 0)
+        return 0;
+
+    if (ctx.count > max_len)
+        ctx.count = max_len;
+
+    // ITER_UBUF only exists in kernels >= 6.0 - earlier kernels use ITER_IOVEC
+    if (LINUX_KERNEL_VERSION >= KERNEL_VERSION(6, 0, 0) && ctx.iter_type == iter_ubuf) {
+        return bpf_probe_read(buf, ctx.count, ctx.ubuf) == 0 ? ctx.count : 0;
     }
 
-    struct iovec vec;
-    bpf_probe_read(&vec, sizeof(struct iovec), iov);
+    if ((ctx.iter_type & ITER_IOVEC) != ITER_IOVEC) {
+        return 0;
+    }
 
-    bpf_dbg_printk("standard iov %llx base %llx len %d", iov, vec.iov_base, vec.iov_len);
+    bpf_clamp_umax(ctx.nr_segs, 4);
 
     u32 tot_len = 0;
 
     // Loop couple of times reading the various io_vecs
-    for (int i = 0; i < 4; i++) {
-        void *p = &iov[i];
-        bpf_probe_read(&vec, sizeof(struct iovec), p);
-        // No prints in loops on 5.10
```

Comment:
Older kernels don't support `bpf_printk` with more than 3 arguments. Splitting the statement does the trick.